lib/propolis/src/firmware/smbios/table.rs

```diff
@@ -46,14 +68,303 @@ impl Table for Type0 {
     }
 }
 
+macro_rules! serialize_enums {
+    ($($Enum:ty => $repr:ty),+ $(,)?) => {
+        $(
+            #[automatically_derived]
+            impl<'de> Deserialize<'de> for $Enum {
+                fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+                where
+                    D: de::Deserializer<'de>,
+                    $repr: Deserialize<'de>,
+                {
+                    lazy_static::lazy_static! {
+                        static ref ERR_MSG: String = {
+                            use std::fmt::Write;
+                            let mut err = "expected one of: ".to_string();
+                            let mut variants = <$Enum>::VARIANTS.iter().copied().peekable();
+                            let mut first = true;
+                            while let Some(variant) = variants.next() {
+                                let has_remaining = variants.peek().is_some();
+                                let or = if !has_remaining && !first {
+                                    " or "
+                                } else {
+                                    ""
+                                };
+
+                                let comma = if has_remaining {
+                                    ", "
+                                } else {
+                                    ""
+                                };
+                                write!(err, "{or}{variant:?} ({:#04x}){comma}", variant as $repr).unwrap();
+                                first = false;
+                            }
```

Comment:
Might be simpler to impl this in terms of `slice::join`, switching on `VARIANTS.len()` if you want to preserve the fancy behavior for the 1 and 2 item cases.

Up to you.